Code/Framework/AzCore/AzCore/DOM/DomPrefixTree.inl

```diff
@@ -0,0 +1,222 @@
+
+                        for (const auto& entry : value->m_values)
+                        {
+                            // The stack runs this in reverse order, so we'll:
+                            // 1) Push the current path entry to currentPath
+                            // 2a) Process the value at the path (if any)
+                            // 2b) Process the value's descendants at the path (if any)
+                            // 3) Pop the path entry from the stack
+                            stack.push(PopPathEntry{});
+                            stack.push(&entry.second);
+                            stack.push(entry.first);
+                        }
+                    }
+                    else if constexpr (AZStd::is_same_v<CurrentType, PathEntry>)
+                    {
+                        currentPath.Push(value);
+                    }
+                    else if constexpr (AZStd::is_same_v<CurrentType, PopPathEntry>)
+                    {
+                        currentPath.Pop();
+                    }
+                },
+                entry);
+        }
+    }
+
+    template<class T>
+    T* DomPrefixTree<T>::ValueAtPath(const Path& path, PrefixTreeMatch match)
+    {
+        // Just look up the node if we're looking for an exact path
+        if (match == PrefixTreeMatch::ExactPath)
+        {
+            if (Node* node = GetNodeForPath(path); node != nullptr && node->m_data.has_value())
+            {
+                return &node->m_data.value();
+            }
+            return {};
+        }
+
+        // Otherwise, walk to find the closest anscestor with a value
+        Node* node = &m_rootNode;
+        T* result = nullptr;
+        const size_t lengthToIterate = match == PrefixTreeMatch::SubpathsOnly ? path.Size() - 1 : path.Size();
+        for (size_t i = 0; i < lengthToIterate; ++i)
+        {
+            if (node->m_data.has_value())
+            {
+                result = &node->m_data.value();
+            }
+
+            const PathEntry& entry = path[i];
+            auto entryIt = node->m_values.find(entry);
+            if (entryIt == node->m_values.end())
+            {
+                break;
+            }
+
+            node = &entryIt->second;
+        }
+
+        if (node->m_data.has_value())
+        {
+            result = &node->m_data.value();
+        }
+
+        return result;
+    }
+
+    template <class T>
+    const T* DomPrefixTree<T>::ValueAtPath(const Path& path, PrefixTreeMatch match) const
+    {
+        // Const coerce the ValueAtPath result, which doesn't mutate but returns a mutable pointer
+        return const_cast<DomPrefixTree<T>*>(this)->ValueAtPath(path, match);
+    }
+
+    template<class T>
+    T DomPrefixTree<T>::ValueAtPathOrDefault(const Path& path, const T& defaultValue, PrefixTreeMatch match) const
+    {
+        const T* value = ValueAtPath(path, match);
+        return value == nullptr ? defaultValue : *value;
+    }
```

Comment:
This is forcing a copy of the default value.

This should use a forwarding reference of `T&& defaultValue` and return `AZStd::forward<T>(defaultValue)` to make sure an unneccessary copy is performed when the move operator is available

Reply:
Done, cheers.